Add whippet

arnaldo2792 opened the pull request:
**Issue number:**

Part of #660

**Description of changes:**

This adds `whippet`, a minimal implementation of a dbus-launcher. The set of features implemented by whippet include:
- Systemd socket activation
- Policy parsing but using TOML

<details>

**Differences between the dbus launcher and whippet**

**No default connect rule for whippet’s user**

One of the key differences between the launcher and whippet is that whippet doesn’t insert a default connect rule, the reason for this is twofold:

* Whippet doesn’t have to interact with the bus, as it doesn’t support the features that require interactions with it
    * Dbus service activation
    * Reload configurations
* Bottlerocket’s default dbus policy allows all system users to connect to the bus
    * So instead of allowing whippet to connect to the bus by default, it only gets the permission if actually allowed by the policy

**Drop support for deprecated or unused rule configurations**

To keep whippet’s implementation as minimal as possible whippet only supports the rule fields that are actually used in Bottlerocket. Additionally, whippet removes fields marked as deprecated in the configuration file and policy rules, this includes

* In configuration file:
    * user: used to drop privileges to the selected user. whippet relies on systemd directives to drop this privilege.
    * auth: [this isn’t implemented](https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/config.c#L1173) in the dbus-launcher
    * listen: [this isn’t implemented](https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/config.c#L1170) in the dbus-launcher
* In Rules:
    * send_type/receive_type = method_return or error: the dbus-launcher [warns and ignores these values](https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/policy.c#L439)
    * Don’t support group, at_console, no_console or mandatory contexts, as they aren’t currently used in Bottlerocket

**Rule priority assignment**

In the dbus-launcher, rules are given a priority that is used to determine which rule wins when sending or receiving a message through the bus. The rules with the highest priorities win.

Whippet implements the priority assignment as close as possible to what the dbus-launcher does. However, there is a limitation on whippet’s implementation. The dbus-launcher assigns priorities to the rules in the order they appear, e.g. these rules:

```xml
<policy context="default">
    <allow user="*" />
    <deny send_path="*" />
</policy>

<policy user="root">
    <allow send_path="*" />
</policy>
```

Will get different priorities than the following rules, even though the actual permissions are identical:

```xml
<policy user="root">
    <allow send_path="*" />
</policy>

<policy context="default">
    <allow user="*" />
    <deny send_path="*" /> 
</policy>
```

This difference in order doesn’t affect the purpose of the policy, as the formula to calculate the rules guarantees that user rules will always get a higher priority to the default rules:

```
PRIORITY_BASE: u64 = u64::MAX / 7
priority = PRIORITY_BASE * <WEIGHT> + *current_priority

WEIGHT FOR DEFAULT = 1 (priority starts at 2635249153387078802)
WEIGHT FOR GROUP = 2 # not used in Bottlerocket
WEIGHT FOR USER = 3 (priority starts at 7905747460161236406)
```

For a default rule to overlap with the next context, there would have to be at least `2635249153387078802` rules. Bottlerocket has 204 rules in total.

**Generated priorities**

After re-arranging the configuration files for both whippet and the dbus-launcher to have the same order, and removing unused rules from the dbus-launcher configuration file (so that they are more deterministic), I confirmed:

|	|Policy size	|Default Policy Rules	|UID 0	|UID 192	|
|---	|---	|---	|---	|---	|
|whippet	|104375 bytes	|Own rules: 1, Send rules: 186, Recv rules: 6	|Own rules: 3, Send rules: 191, Recv rules: 9	|Own rules: 2, Send rules: 187, Recv rules: 7	|
|dbus-launcher	|104375 bytes	|Own rules: 1, Send rules: 186, Recv rules: 6	|Own rules: 3, Send rules: 191, Recv rules: 9	|Own rules: 2, Send rules: 187, Recv rules: 7	|

**Sample rules to demonstrate they got the same priority**

**Launcher**

```
Send rule 13: verdict=ALLOW, priority=2635249153387078828, name='org.freedesktop.login1', path='', interface='org.freedesktop.login1.Manager', member='GetSession',
```

**Whippet**

```
Send rule 13: verdict=ALLOW, priority=2635249153387078828, name='org.freedesktop.login1', path='', interface='org.freedesktop.login1.Manager', member='GetSession',
```

</details>

**NOTE FOR REVIEWERS**

Reviewers are advised to review in this order:
- `main.rs`
- `policy.rs`: contains information of how rules are assigned a priority, and provides examples and unit tests cases for better understanding
- `dbus_policy.rs`: contains information of how the default rules are passed to all the users in a policy
- `broker.rs`
- `child.rs`


**Testing done:**
In combination with #677 and #678: 

<details>

- Kubernetes 1.31, 6.1 Kernel, systemd 252, using whippet:

```bash
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "90a6e5c4f-dirty",
    "pretty_name": "Bottlerocket OS 1.47.0 (aws-k8s-1.31)",
    "variant_id": "aws-k8s-1.31",
    "version_id": "1.47.0"
  }
}
# whippet is running as the dbus-broker service
bash-5.1# systemctl status whippet.service
● whippet.service - D-Bus System Message Bus
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/whippet.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (running) since Fri 2025-10-10 05:06:52 UTC; 10min ago
TriggeredBy: ● dbus.socket
   Main PID: 1165 (whippet)
      Tasks: 6 (limit: 18904)
     Memory: 5.0M
        CPU: 146ms
     CGroup: /system.slice/whippet.service
             ├─1165 /usr/bin/whippet
             └─1170 /usr/bin/dbus-broker --log 12 --controller 11 --machine-id ec205cbda35e60eb68d2ca9ea40b8d61 --max-bytes 536870912 --max-fds 4096 --max-matches 16384

Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] read_commands; n_commands=1
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] Using listener socket FD: 3
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] Calling AddListener method
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] socket reader;
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] read_socket;
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] read_socket;
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] AddListener call completed successfully
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] Policy applied successfully
Oct 10 05:06:52 localhost whippet[1165]: 05:06:52 [INFO] Notified systemd that service is ready
Oct 10 05:06:52 localhost systemd[1]: Started D-Bus System Message Bus.

bash-5.1# systemd-analyze --version
systemd 252 (252)
-PAM -AUDIT +SELINUX -APPARMOR -IMA -SMACK +SECCOMP -GCRYPT -GNUTLS +OPENSSL +ACL +BLKID -CURL -ELFUTILS -FIDO2 -IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK -PCRE2 -PWQUALITY -P11KIT -QRENCODE +TPM2 -BZIP2 -LZ4 -XZ -ZLIB -ZSTD -BPF_FRAMEWORK -XKBCOMMON -UTMP -SYSVINIT default-hierarchy=unified

# The dbus-broker-launcher package isn't installed when whippet is chosen
bash-5.1# grep "dbus-broker-launcher" /usr/share/bottlerocket/application-inventory.json
bash-5.1# grep "whippet" /usr/share/bottlerocket/application-inventory.json
      "Name": "whippet",
bash-5.1#
```
</details>


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
sources: add whippet

whippet is a dbus-broker launcher that implements the minimal set of
featured required by Bottlerocket to get a functional Bus, which
includes:

- D-Bus Policy parsing
- Systemd socket activation

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
selinux-policy: use whippet as a bus_exec entrypoint

whippet requires the bus_exec treatment for it to interact with the
dbus-broker.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

KCSesh commented on packages/selinux-policy/fs.cil:
nit: Typo on commit "selinux-policy: use whippet as a bus_exec entrypoint"
`iteract` -> `interact`

arnaldo2792 commented:
Forced push includes:
- Provide the `dbus-launcher` capability with the lowest priority through `whippet`

bcressey commented on packages/os/os.spec:
```diff
@@ -423,6 +424,13 @@ Conflicts: %{_cross_os}bash
 
+%package -n %{_cross_os}whippet
+Summary: A simple dbus-broker launcher
+Provides: %{_cross_os}dbus-launcher = %{package_priority_epoch}:
+Conflicts: %{_cross_os}dbus-launcher
```
I'd recommend using parentheses for the capabilities that aren't expected to match a package name:
```suggestion
Provides: %{_cross_os}dbus-broker(launcher) = %{package_priority_epoch}:
Conflicts: %{_cross_os}dbus-broker(launcher)
```

bcressey commented on sources/Cargo.lock:
```diff
@@ -3573,10 +3798,23 @@ dependencies = [
  "pin-utils",
 ]
 
+[[package]]
+name = "nix"
+version = "0.30.1"
```
Can we update our pinned workspace version of nix so we don't need an extra version?

bcressey commented on sources/Cargo.lock:
```diff
@@ -559,6 +559,91 @@ dependencies = [
 
+[[package]]
+name = "async-broadcast"
+version = "0.7.2"
```
Can we disable the async feature of zbus to avoid bringing in these additional dependencies? Not sure what async is really buying us in this context.

bcressey commented on sources/Cargo.toml:
```diff
@@ -206,6 +208,8 @@ unindent = "0.2"
 url = "2"
 walkdir = "2.4"
 which = "4"
+zbus = { version = "5.11", features = ["p2p"] }
+zvariant = "5.7"
```
Let's do another pass over these dependencies and make sure we disable any features we don't actually need.

bcressey commented on sources/deny.toml:
```diff
@@ -66,6 +66,9 @@ skip-tree = [
     # aws-smithy-experimental brings a set of new dependencies that
     # the aws-sdk doesn't support
     { name = "aws-smithy-experimental", version = "=0.1" },
+    # zbus introduces newer versions of crates that are either
+    # incompatible or way ahead from the versions used in other crates
+    { name = "zbus", version = "=5.11" }
```
I'd like to try to resolve this now if we can or else document what's keeping us on older crates.

arnaldo2792 replied:
I created an issue to track these:

https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/674

bcressey reviewed (commented):
Initial review covering build-related concerns.

bcressey commented on sources/whippet/README.md:
````diff
@@ -0,0 +1,118 @@
+```
+
+*whippet* prevents the definition of invalid rules by using actual types, to prevent users from mixing
+fields for different rules:
````
I don't really follow the distinction you're drawing here. For the upstream launcher, the canonical policy format is XML, so it must be parsed and invalid constructs must be rejected at runtime. For whippet, the canonical policy format is TOML, so it must be deserialized and invalid constructs will be rejected at runtime.

Those seem essentially the same to me - we can't stop people from writing garbage policies, so we have to identify garbage and throw it out.

I'd prefer to just provide examples of translating valid policies from XML to TOML so it doesn't come off as a language flex.

bcressey commented on sources/whippet/README.md:
```diff
@@ -0,0 +1,118 @@
+    { allow = true, user = "*" },             # This is a connect user rule
+    { allow = true, own = "*" },              # This is an own rule
+    { allow = true, send_destination = "*" }, # This is a send rule
+    { allow = true , receive_sender = "*"},   # This is a receive rule
```
```suggestion
    { allow = true, receive_sender = "*"},   # This is a receive rule
```

bcressey commented on sources/whippet/README.md:
```diff
@@ -0,0 +1,118 @@
+The original dbus-daemon allowed setting the user the daemon should run as. In whippet, the user for
+the bus is configured in the systemd unit file that runs the bus. This allows whippet to only care about
+starting the broker, and keeps its implementation simpler as it doesn't have to implement any sort of
+privilege-dropping mechanism.
```
Here you switch from talking about "the original dbus-daemon" to "the broker". I'd just stick to summarizing this implementation's key points, like "doesn't change users, runs with whatever user is configured in the systemd unit", since those are more important to integrators than a comparison with other implementations.

This also buries a potential security concern that should be made clear: unlike other implementations, if you start this launcher as root with full capabilities, that's what the broker will end up running with too.

bcressey commented on sources/whippet/README.md:
```diff
@@ -0,0 +1,118 @@
+
+#### Policy drop-in files
+
+The original dbus-daemon implmentation allows system services to provide their own custom policy
```
nit: probably worth a spell-checker pass
```suggestion
The original dbus-daemon implementation allows system services to provide their own custom policy
```

bcressey commented on sources/whippet/README.md:
```diff
@@ -0,0 +1,118 @@
+
+*whippet* implements the priority assignment as close as possible to what the dbus-launcher does. The dbus-launcher
+numbers the rules in a policy as a whole, considering all contexts in the order they appear In these policies, the
+same rules end up with a different index and therefore a different priority:
```
Who is your audience for this note? Policy authors? Dbus implementation experts?

It's confusing since I'm not sure it's actionable. You say "whippet implements the priority assignment as close as possible to what the dbus-launcher does" here and then go on to explain how it's different. But it's not clear how that difference matters in practice. Also, "as close as possible" implies "identical" unless you explain why it's not possible to mimic the behavior exactly.

Here again I think the documentation would be better written from the standpoint of what policy authors need to do (or not do) or need to be aware of, or what isn't supported that they might expect.

bcressey commented on sources/whippet/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
```
You should add a copyright header for the dbus-broker project here also using information from the AUTHORS doc.

bcressey commented on sources/whippet/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+    // The listener socket has to be configured as O_NONBLOCK, otherwise clients that connect to
+    // the dbus socket will hang
+    let current_flags =
+        nix::fcntl::fcntl(fd, nix::fcntl::FcntlArg::F_GETFL).context(error::GetFDFlagsSnafu)?;
```
nit: I'd add use statements for `fcntl::fcntl` and the `FcntlArg` constants to make this more readable

bcressey commented on sources/whippet/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+///
+/// Validates that exactly one socket was provided via systemd socket activation
+/// and configures it with the required flags for dbus-broker operation.
+#[must_use = "listener socket file descriptor must be used"]
```
I don't really follow the need for these `must_use` markers. I don't think it's necessarily a program bug to call these and throw away the resulting sockets.

bcressey commented on sources/whippet/src/socket.rs:
```diff
@@ -0,0 +1,139 @@
+use snafu::{ensure, ResultExt};
+use std::env;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::os::unix::net::{UnixDatagram, UnixStream};
```
Generally it's an error to use the blocking sockets from std in an async program, so if we need `async` then you should use the Tokio versions.

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+    #[argh(
+        option,
+        long = "config-file",
+        default = "String::from(\"/usr/share/dbus-1/system.toml\")"
```
Let's not put non-standard config files in the standard `/usr/share/dbus-1` path:
```suggestion
        default = "String::from(\"/usr/share/whippet/system.toml\")"
```

ginglis13 commented on sources/Cargo.toml:
```diff
@@ -77,6 +77,8 @@ members = [
     "constants",
 
     "xfscli",
+
+    "whippet",
```
nit - it looks like this list is meant to be in alphabetical order (with `constants` as an outlier...)

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    ReadPolicyFile {
+        path: String,
+        source: std::io::Error,
+    },
```
These could all be consolidated into a `ReadPath` error that takes a `what` argument in addition to the path.

ginglis13 commented on sources/whippet/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "whippet"
+version = "0.1.0"
```
You can give yourself some credit here 😄 

```suggestion
version = "0.1.0"
authors = [ "..." ]
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    MissingListenFds { source: std::env::VarError },
+
+    #[snafu(display("Failed to retrieve NOTIFY_SOCKET environment variable"))]
+    MissingNotifySocket { source: std::env::VarError },
```
These could all be `MissingEnvVar` errors that take a "what" parameter.

ginglis13 reviewed (commented):
Some initial feedback before I dive deeper into the code

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    #[snafu(display("LISTEN_PID mismatch, '{expected}', found {found}"))]
+    UnexpectedPid { expected: String, found: String },
+
+    #[snafu(display("Expected at most 1 file descriptor, foudn {found}"))]
```
```suggestion
    #[snafu(display("Expected at most 1 file descriptor, found {found}"))]
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    },
+
+    #[snafu(display("Failed to get FD flags"))]
+    GetFDFlags { source: nix::errno::Errno },
```
nit: might also be helpful to log which file descriptor this fails for
```suggestion
    #[snafu(display("Failed to get file descriptor flags"))]
    GetFdFlags { source: nix::errno::Errno },
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    JournalSocket { source: std::io::Error },
+
+    #[snafu(display("Error while creating socket for systemd notification: {source}"))]
+    SystemdNotifySocket { source: std::io::Error },
```
nit: I think we don't generally single-quote `{source}` but let's be consistent with other programs and also within this file

arnaldo2792 replied:
They aren't actually needed since I'm using the `report` API in snafu, so I'll clean them.

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+    ClearCloexecFlag { source: nix::errno::Errno },
+
+    #[snafu(display("Failed to clear NONBLOCK flag for socket"))]
+    ClearNonBlockFlag { source: nix::errno::Errno },
```
nit: `ClearFlag` error that takes a "what"

bcressey commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,125 @@
+        debug!("Setting up signal handlers");
+        let mut sigterm = signal::unix::signal(signal::unix::SignalKind::terminate())
+            .context(error::CreateSigTermSignalSnafu)?;
+        let child = self.child.as_mut().context(error::ChildNotRunningSnafu)?;
```
The error message you added for `ChildNotRunningSnafu` - "Called wait in a child that hasn't started" - implies that this is something that should be retried, not a fatal error.

Also, unless I've misunderstood the code flow, we can't be "in" a child here.

You might want to look into the typestate pattern so that we have distinct `BrokerChild` types to reflect whether we've spawned the child or not. Then we wouldn't have to address the case where `wait` is called before `run` because the type system wouldn't allow it.

bcressey commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,125 @@
+                Ok(())
+            }
+            // TODO: handle the kill signal in the the broker, and send the signal down through
+            // a channel
```
If it's not worth doing now, I wouldn't add a TODO for it.

bcressey commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,125 @@
+            })?
+            .trim()
+            .to_string();
+        info!("Machine ID: {}", machine_id);
```
I wouldn't elevate this to `info`:
```suggestion
        debug!("Machine ID: {}", machine_id);
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,135 @@
+        record_type: String,
+    },
+
+    #[snafu(display("Cannot convert UID '{uid}' to integeger"))]
```
```suggestion
    #[snafu(display("Cannot convert UID '{uid}' to integer"))]
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        debug!("Creating BrokerManager with broker path: {}", broker_path);
+
+        debug!("Creating controller socket pair");
+        let (controller, broker_end) = socket::controller_pair()?;
```
maybe `launcher_socket` and `child_socket` or `broker_socket` would be more clear.

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        let journal_socket = socket::journal_socket()?;
+
+        // Transfer ownwership of the sockets to the child process, so that they are cleaned up
+        // automatically once the child goes out of scope (no extra drop calls required)
```
```suggestion
        // Transfer ownership of the sockets to the child process, so that they are cleaned up
        // automatically once the child goes out of scope.
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+    pub(crate) async fn run(&mut self) -> Result<()> {
+        // Start the broker in the background but don't block on it being ready, as the AddListener
+        // call will block until the broker is ready to receive calls in the controller socket
+        self.child.run()?;
```
I don't understand the "don't block on it being ready" comment:
```suggestion
        self.child.run()?;
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        // call will block until the broker is ready to receive calls in the controller socket
+        self.child.run()?;
+        debug!("Adding listener with parsed policy");
+        self.add_listener().await?;
```
the "add_listener" method name really doesn't capture what's going on - a listener sounds passive. Maybe one of these:
```suggestion
        self.apply_policy().await?;
        self.set_listener_policy().await?;
        self.set_child_policy().await?;
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        info!("Policy applied successfully");
+        // Notify the broker is ready once the policy and the sockets were passed to the broker.
+        // This deviates slightly from the dbus-launcher as it notifies systemd right after the
+        // configuration is loaded
```
I think the actual behavior is obvious from code flow, and I don't think it matters what the launcher does:
```suggestion
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        self.notify_systemd_ready()?;
+        info!("Notified systemd that service is ready");
+
+        // Wait forenver until signaled to stop or the dbus-broker is killed
```
nit: or just omit this comment too
```suggestion
        // Wait forever until signaled to stop or the dbus-broker is killed
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        info!("Calling AddListener method");
+
+        // Wrap policy in variant (like launcher does with sd_bus_message_open_container(m, 'v', ...))
+        let policy_variant = Value::from(self.policy.clone());
```
Let's import `Value` as `ZVariantValue` or `ZValue` or something more distinct.

bcressey replied:
nit: avoid including type information in the variable name:
```suggestion
        let policy = Value::from(self.policy.clone());
```

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+        proxy
+            .add_listener(object_path, fd_handle, policy_variant)
+            .await
+            .context(error::DbusBrokerAddListenerSnafu)?;
```
I believe you said that we block until the policy is retrieved and loaded by the child, but it's not obvious where or why that happens. Can you add more comments to explain what's happening behind the scenes?

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,138 @@
+)]
+trait DBusBroker {
+    /// Implements the AddListener call to transfer the systemd socket and dbus policy to the
+    /// broker
```
This comment confuses me - it's really the broker that implements this, and the launcher just ends up calling this IPC method, right? So this trait is more like a list of known function signatures available via FFI than an implementation.

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,434 @@
+/// Top-level policy structure that matches launcher's Dbus format. It is crucial that
+/// the order of the fields remains like it is, otherwise the broker rejects the policy.
+///
+/// Use zvariant's Type and Value to generate both the Dbus signature (avaialable at
```
```suggestion
/// Use zvariant's Type and Value to generate both the Dbus signature (available at
```

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,434 @@
+//! D-Bus policy serialization and conversion.
```
Need to add the copyright and attribution information here since you're referencing dbus-broker sources and implementation details so heavily.

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,434 @@
+    pub(crate) recv_rules: Vec<SendReceiveRecord>,
+}
+
+/// Represents an Own Record in the actual dbus policy
+/// See:
+/// https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/policy.c#L836
```
If possible, it'd be better to link to the broker side of this interface for any of these data structures - i.e. a `.h` rather than a `.c` file.

arnaldo2792 replied:
I could, but the launcher uses the same type to represent all the serialized records:

https://github.com/bus1/dbus-broker/blob/a7960f21977059dbb2356072bcd6f583b667593c/src/launch/policy.h#L21

I could add both tho, but I thought the link I provided first will be more meaningful to the implementation as that's what actually dictates the form of the serialized record.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+use std::path::Path;
+
+/// Default directory path additional policies are read from
+pub(crate) const DBUS_POLICY_DROPINS_PATH: &str = "/usr/share/dbus-1/system.d";
```
```suggestion
/// Default directory path additional policies are read from
pub(crate) const DBUS_POLICY_DROPINS_PATH: &str = "/usr/share/whippet/settings.d";
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+/// Default directory path additional policies are read from
+pub(crate) const DBUS_POLICY_DROPINS_PATH: &str = "/usr/share/dbus-1/system.d";
+/// Multiplier used to calculate the priority of a rule
+/// https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/policy.c#L237
+const DEFAULT_POLICY_CONTEXT_MULTIPLIER: u64 = 1;
+const DEFAULT_POLICY_CONTEXT_USER: u64 = 3;
+const PRIORITY_BASE: u64 = u64::MAX / 7;
```
The constants themselves aren't copyrightable, but the algorithm you're using that depends on them comes from dbus-broker, so you should attribute this in a copyright header.

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,434 @@
+    //
+    // record_type is the counter part of the policy's MessageType, so at least that enum will
+    // enforce only valid values for record_type
+    pub record_type: u32,
```
Looks like you can use `serde_repr` for this, which is part of the `zbus` dependency graph so it's not a new dependency.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+const PRIORITY_BASE: u64 = u64::MAX / 7;
+
+/// Represents a minimal policy configuration for the dbus-broker. Each section on the policy is
+/// known as a Context.
+#[derive(Debug, Deserialize, Default)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct Policy {
+    /// The default rules that apply to all contexts. This maps to the original <policy
+    /// context="default">
+    pub(crate) default: Option<Context>,
+    /// User-specific rules. This maps to the original <policy user="<user>">
+    pub(crate) user: Option<IndexMap<String, Context>>,
+}
```
These comments aren't very meaningful. Why is each section on the policy known as a `Context`? Does that matter? Why do I need to know what the struct fields map to in terms of XML fragments?

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+    pub(crate) rules: Vec<Rule>,
+}
+
+/// Rules that are allowed in Whippet's implementation
```
Better would be to summarize what a rule is

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+    // The original dbus policy defines send_requested_reply, but this field is not used in the
+    // dbus-launcher in an impactful way, it is only used to print some warnings. The value isn't
+    // serialized in the final policy.
+    Send {
```
Is this comment about `Send` rules and what they do, or about an unimplemented `SendRequestedReply` rule variant? If it's the latter, I would just delete this and summarize what a `Send` rule is.

(Same comment about `Receive`.)

arnaldo2792 replied:
The former. I will update the main documentation to explain this, and just comment here what a send rule is like in the other rules.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+
+        let mut policy: Self = toml::from_str(&policy_content).context(error::PolicyParseSnafu)?;
+
+        // The first rule in processed by the launcher actually gets priority 3 because:
+        // - the dbus-launcher starts its priority counter at 1
+        // - the dbus-launcher inserts a default connect rule for the current user with priority 2
+        // Hence, use the same priority as the actual first rule that will be processed
```
Does `whippet` do the same - insert a default connect rule for the current user with priority 2?

arnaldo2792 replied:
It doesn't, but it could. I avoided that because whippet doesn't really need it. The default connect rule for the current user is needed because the dbus-broker-launcher actually connects to the bus through the `dbus` socket to support service activation. Since whippet doesn't support this feature, I thought the connect rule wasn't needed. Plus, the default policy in Bottlerocket allows all users to connect to the bus so ultimately the default rule would have been kept.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+    }
+
+    /// Validates the rules of the policy according to the restrictions set by the dbus-launcher
+    /// - No connect rules for the user contexts
```
nit: a list with one element seems strange:
```suggestion
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+
+/// Resolves the provided username string to its corresponding UID
+fn resolve_username_to_uid(username: &str) -> Result<u32> {
+    debug!("Resolving username '{}' to UID", username);
```
It'd be useful to memoize resolved usernames, as this could be expensive

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+        .into_iter()
+        .map(|entry| entry.path())
+        .filter(|path| path.extension().and_then(|s| s.to_str()) == Some("toml"))
+        .collect();
```
nit: could just use `walkdir`

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+        })?;
+
+        let mut policy: Policy = toml::from_str(&content).context(error::PolicyParseSnafu)?;
+        policy.set_rule_priorities(&mut current_priority);
```
You can factor this part into a `load_one_policy` helper, and use it in `new()` also.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+            } else {
+                main_policy.user = Some(user_rules);
+            }
+        }
```
if `load_one_policy` took a base policy, then you could call it with `Policy::default()` in `new()`, and move this logic into it as well.

Overall I'd like make the logic `load_additional_policies` look like as much like `new()` as possible. You could also merge them and rename `load_additional_policies` to just `load_policies`.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+            MessageType::Signal => 4,
+        }
+    }
+}
```
`serde_repr` might let you get rid of this also

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+                let uid = resolve_username_to_uid(&user)?.to_string();
+                let policy = users_contexts.shift_remove(&user).unwrap_or_default();
+                users_contexts.insert(uid, policy);
+            }
```
This isn't very efficient. Instead I'd suggest draining the original `users_contexts` and creating a new IndexMap, replacing the username with the resolved UID and the original rules.

bcressey commented on sources/whippet/src/policy.rs:
````diff
@@ -0,0 +1,861 @@
+    /// ]
+    /// ```
+    ///
+    /// After this funciton is called, the result is:
````
```suggestion
    /// After this function is called, the result is:
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+            self.user = Some(IndexMap::new());
+        }
+        // Unlikely to get the error as the line above guarantees the user policy should be set
+        let user_policy = self.user.as_mut().expect("Unexpected empty user context");
```
Why have this comment - is it "unlikely" or impossible?

arnaldo2792 replied:
I put the comment just to explain that the `expect` here doesn't do any harm, and it should be impossible for it to cause a panic. I looked at other places in Bottlerocket to see what was done before with this type of calls, and no comment was added to explain them so I'll remove the comment.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+                .collect();
+
+            connect_rules.sort_by(|a, b| b.get_priority().cmp(a.get_priority()));
+            let default_priority = connect_rules.first().map(|r| *r.get_priority());
```
`priority_floor` or `priority_threshold` might be a better name - "default" is somewhat overloaded and confusing.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+                            true
+                        } else {
+                            false
+                        }
```
Here we end up using a user's connect rule as the new priority threshold if it's greater than our initial threshold.

However, if we don't have an initial threshold, we don't enter this loop at all. That seems inconsistent, maybe incorrect. Why should we discard subsequent lower priority user rules in the first case but keep them unconditionally in the second case?

arnaldo2792 replied:
Thanks for catching this! There was indeed a bug.

ginglis13 commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+# Introduction
+
+*whippet* is Bottlerocket's implementation of the dbus-launcher. It implements the minimal
+set of features to configure and start the `dbus-broker`, which incude:
```
nit: 
```suggestion
set of features to configure and start the `dbus-broker`, which include:
```

ginglis13 commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,187 @@
+    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::LoggerSnafu)?;
+
+    info!("Starting whippet D-Bus launcher");
+    debug!("Config file: {}", args.config_file);
+    debug!("Broker path: {}", args.broker_path);
```
should we validate these paths exists before continuing?

arnaldo2792 replied:
The `new` function will validate if they exist, and error out if they don't.

ginglis13 commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+#[derive(Debug, Deserialize, Clone, PartialEq)]
+#[serde(untagged, deny_unknown_fields)]
+pub(crate) enum Rule {
+    // Connect users dictates whether a user is allowed to connect to the bus
```
For each of the `Rule` variants you can provide these comments as docstrings:

```suggestion
    /// Connect users dictates whether a user is allowed to connect to the bus
```

ginglis13 commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+                .extract_if(.., |rule| {
+                    // Extract ONLY connect rules, that aren't for the DEFAULT user
+                    if let Rule::ConnectUser { user, .. } = rule {
+                        *user != u32::MAX.to_string()
```
there are multiple places in this PR where it's implied that `u32::MAX` is the default user. It might be better to define a constant
```
const DEFAULT_USER_ID: u32 = u32::MAX
```

ginglis13 commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+        // Unlikely to get the error as the line above guarantees the user policy should be set
+        let user_policy = self.user.as_mut().expect("Unexpected empty user context");
+
+        // Iterate over the connect_rules found for specific users, inserting a new user policy
+        // when it doesn't exist.
+        while let Some(rule) = connect_rules.pop() {
```
Does the ordering matter here since `.pop` pops from the back of the connect_rules? Or is that ordering handled by the rule's priority?

arnaldo2792 replied:
Ordering doesn't matter here, it only matters while setting the priorities.

ginglis13 commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,861 @@
+                });
+            }
+
+            // Remove any user contexts without any rules
+            user_policy.retain(|_, p| !p.rules.is_empty());
```
would it be technically more efficient to do this first before the priority comparison above?

arnaldo2792 replied:
We have to do this at the end because the priority comparison could drain all the rules.

ginglis13 commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,434 @@
+    pub(crate) bus_type: String,
+}
+
+/// Represents the actual policy in the dbus launcher, as with the DbusPolicy, the order of the
```
```suggestion
/// Represents the actual policy in the dbus-launcher, as with the DbusPolicy, the order of the
```

ginglis13 commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,125 @@
+const DEFAULT_MAX_CONNECTIONS_PER_USER: u64 = 64;
+const DEFAULT_MAX_MATCH_RULES_PER_CONNECTION: u64 = 256;
+
+/// Calculated limits (per-user = per-connection × max-connections)
```
```suggestion
/// Calculated limits (per-user = per-connection × max-connections)
/// See: https://github.com/bus1/dbus-broker/blob/b0db0890d1254477cf832e5f9f0a798360c80fd9/src/launch/launcher.c#L1102-L1105
```

arnaldo2792 commented:
Forced push includes:

- Added COPYRIGHT headers to the files that required it
- Drop `async-io` in favor of `tokio`, this required refactoring how the socket handles are passed to other structs, as they don't have to be drop (closed), otherwise the broker stops
- Refactor to use the `typestate` pattern as requested
- Provide `dbus-broker(launcher)`
- Fixed bug in how the default user rules are optimized
- Refactor `load_additional_policies` as requested
- Extra check on dropping additional features from `zbus` and `zvariant`
- Provide `drop-in` for the dbus-broker to use `whippet` as the `ExecStart` command
- Condensed error types to be more re-usable

arnaldo2792 commented:
(Forced push includes rebase)

arnaldo2792 commented:
(Forced push fixes clippy failures)

ginglis13 commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+        // easier for debugging
+        dbus_policy
+            .uid_entries
+            .insert(0, (u32::MAX, default_batch.clone()));
```
```suggestion
            .insert(0, (DEFAULT_USER_ID, default_batch.clone()));
```

ginglis13 commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+                {{ allow = true, send_interface = "org.example.Dummy" }}
+            ]
+        "#,
+            u32::MAX
```
```suggestion
            DEFAULT_USER_ID
```

ginglis13 commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+                {{ allow = true, send_interface = "org.example.Dummy" }}    # Rule 4
+            ]
+        "#,
+            u32::MAX
```
```suggestion
            DEFAULT_USER_ID
```

ginglis13 commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+        policy.optimize();
+
+        let dbus_policy: DbusPolicy = policy.try_into().unwrap();
+        assert_eq!(dbus_policy.uid_entries[0].0, u32::MAX);
```
```suggestion
        assert_eq!(dbus_policy.uid_entries[0].0, DEFAULT_USER_ID);
```

bcressey commented on packages/os/os.spec:
```diff
@@ -423,6 +425,13 @@ Conflicts: %{_cross_os}bash
 
+%package -n %{_cross_os}whippet
+Summary: A simple dbus-broker launcher
+Provides: %{_cross_os}dbus-broker(launcher) = 0
```
The `:` denotes epoch, otherwise this is providing `Version: 0` for this capability.
```suggestion
Provides: %{_cross_os}dbus-broker(launcher) = 0:
```

bcressey commented on packages/os/os.spec:
```diff
@@ -423,6 +425,13 @@ Conflicts: %{_cross_os}bash
 %{summary}.
 
+%package -n %{_cross_os}whippet
+Summary: A simple dbus-broker launcher
```
nit: let's avoid "simple" as it's overly subjective
```suggestion
Summary: Custom launcher for the D-BUS message broker
```

bcressey commented on packages/os/whippet.conf:
```diff
@@ -0,0 +1,4 @@
+[Service]
+User=dbus
+ExecStart=
+ExecStart=/usr/bin/whippet
```
This should be a freestanding unit and not a drop-in.

bcressey commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,40 @@
+    # This is a userless rule, it is dropped from the rules at runtime
+    # { receive_type = "method-return", allow = true },
+    # This is a userless rule, it is dropped from the rules at runtime
+    # { receive_type = "error", allow = true },
```
Let's just remove the useless rules.

bcressey commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,40 @@
+    # Allow monitoring
+    { send_destination = "org.freedesktop.DBus", send_interface = "org.freedesktop.DBus.Monitoring", allow = true },
+    # Allow stats interface
+    { send_destination = "org.freedesktop.DBus", send_interface = "org.freedesktop.DBus.Debug.Stats", allow = true }
```
Most of these comments aren't adding much value.

I'm OK with just removing them, but if you want to keep them, a comment on *why* this is the right policy would be better than *what* the rule does.

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,132 @@
+    /// Spawn the broker and transfer the systemd socket and policy
+    pub(crate) async fn run(self) -> Result<()> {
+        let journal_socket = socket::journal_socket()?;
+        // The broker has to be started on the background before the connection is created
+        // otherwise the connection hangs because there is nothing connected on the other side
+        // of the socket pair.
+        let running = self.child.run(journal_socket, self.broker_socket)?;
```
If the child process is spawned successfully but later dies, this comment implies we will block forever in one of the subsequent steps - connection builder? set listener policy? - before we actually wait on the child process and discover that it's terminated.

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,132 @@
+        ObjectPath::try_from("/org/bus1/DBus/Listener/0").context(error::ParseObjectPathSnafu)?;
+    // The listener file descriptor technically isn't owned by whippet. Wrap the file
+    // descriptor around a BorrowedFd to prevent the file descriptor from being closed once the
+    // broker goes out of scope
```
This comment is confusing - who owns the file descriptor if not whippet? What happens if it's closed?

bcressey commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,132 @@
+
+    info!("Calling AddListener method");
+
+    // Wrap policy in variant (like launcher does with sd_bus_message_open_container(m, 'v', ...))
```
This comment is meaningless without detailed knowledge of a different implementation. It's also implied by the signature of the AddListener method that we need to send a `ZVariant` type.

bcressey commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,128 @@
+        Ok(Self {
+            broker_path: broker_path.to_string(),
+        })
+    }
```
There aren't any provided sockets?

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,135 @@
+/*!
+# Introduction
```
nit: the line lengths are very inconsistent in this docstring. I'd recommend joining all the lines in sentences and reformatting.

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,135 @@
+    { allow = true, user = "*" },             # This is a connect user rule
+    { allow = true, own = "*" },              # This is an own rule
+    { allow = true, send_destination = "*" }, # This is a send rule
+    { allow = true , receive_sender = "*"},   # This is a receive rule
```
```suggestion
    { allow = true, receive_sender = "*"},   # This is a receive rule
```

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,135 @@
+/*!
+# Introduction
+
+*whippet* is Bottlerocket's implementation of the dbus-launcher. It implements the minimal
```
`dbus-launcher` is not the name of the program you're replacing.
```suggestion
*whippet* is Bottlerocket's implementation of a launcher for `dbus-broker`. It implements the minimal
```

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,135 @@
+- Policy configuration
+- Systemd socket activation
+
+## Writing Dbus policies for whippet
```
Please stick to a consistent set of standard terms and acronyms:
```suggestion
## Writing D-Bus policies for whippet
```

bcressey commented on sources/whippet/src/main.rs:
````diff
@@ -0,0 +1,135 @@
+```
+
+As with the original dbus-daemon policy, the rules in the default context will
+be included in the list of rules that apply to all the users defined in the policy.
````
```suggestion
The rules in the default context will be included in the list of rules that apply to all the users defined in the policy.
```

bcressey commented on sources/whippet/src/socket.rs:
```diff
@@ -0,0 +1,159 @@
+            fd,
+            what: "O_NONBLOCK",
+        },
+    )?;
```
nit: for consistency, I would factor this into a `set_nonblock` helper. Also, `ClearFlagSnafu` is misleading since you are setting (not clearing) the flag.

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+pub enum Error {
+    // Configuration errors
+    #[snafu(display("Failed to initialize logger"))]
+    Logger { source: log::SetLoggerError },
```
We typically emit `source` as part of the display message. If I understand correctly, we don't need to do that here because we have `#[snafu::report]` applied to main?

arnaldo2792 replied:
Correct, none of the `source` are needed because of the `#[snafu::report]` applied to main.

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+    },
+
+    #[snafu(display("User context includes connect rule"))]
+    UserContextWithConnectRule,
```
Can we extend this with the specific user ID / username from the rule so it's easier to find the bad context?

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+    ControllerClone { source: std::io::Error },
+
+    // Socket errors
+    #[snafu(display("Failed to retrieve {what} environment variable"))]
```
```suggestion
    #[snafu(display("Failed to retrieve '{what}' environment variable"))]
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+        source: std::env::VarError,
+    },
+
+    #[snafu(display("LISTEN_PID mismatch, '{expected}', found {found}"))]
```
```suggestion
    #[snafu(display("LISTEN_PID mismatch, '{expected}', found '{found}'"))]
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+        source: std::num::ParseIntError,
+    },
+
+    #[snafu(display("Failed to get FD flags for file descriptor '{fd}'"))]
```
```suggestion
    #[snafu(display("Failed to get flags for file descriptor '{fd}'"))]
```

bcressey commented on sources/whippet/src/error.rs:
```diff
@@ -0,0 +1,130 @@
+    SystemdNotifySocket { source: std::io::Error },
+
+    // Child errors
+    #[snafu(display("Failed to read machine-id from '{path}'"))]
```
```suggestion
    #[snafu(display("Failed to read machine ID from '{path}'"))]
```

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+            PolicyBatch {
+                connect_verdict: false,
+                connect_priority: 1,
+                own_rules: Vec::new(),
+                send_rules: Vec::new(),
+                recv_rules: Vec::new(),
+            }
```
nit: leverage the default impl to make it clear what's actually changing
```suggestion
            let policy_batch = Default::default();
            policy_batch.connect_priority = 1;
            policy_batch
```

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,446 @@
+            .insert(0, (u32::MAX, default_batch.clone()));
+
+        if let Some(user_policies) = &policy.user {
+            for (uid, user_policy) in user_policies {
```
```suggestion
        for (uid, user_policy) in &policy.user.iter().flatten() {
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+        let default_rules = policy.default.unwrap().rules;
+        let user_context = policy.user.unwrap();
+        let alice_rules = &user_context.get(&ALICE_USER.to_string()).unwrap().rules;
+        println!("{alice_rules:?}");
```
nit: remove debugging line
```suggestion
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+
+        // Load all the additional policies that are found in the policies_dir so that all
+        // the users are validated while constructing the config object to stop as soon as possible
+        // if there is an invalid user.
```
This comment should be above the for loop?

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+                final_users_contexts.insert(uid, policy);
+            }
+            self.user = Some(final_users_contexts);
+        }
```
Try something like this to avoid all the shifting and cloning:
```suggestion
        let users_contexts = self.user.take().unwrap_or_default();
        let mut final_users_contexts = IndexMap::new();
        for (user, policy) in users_contexts.into_iter() {
            let uid = username_resolver.resolve(&user)?.to_string();
            final_users_contexts.insert(uid, policy);
        }
        self.user = Some(final_users_contexts);
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+        if self.user.is_none() {
+            self.user = Some(IndexMap::new());
+        }
+        let user_policy = self.user.as_mut().expect("Unexpected empty user context");
```
You could reuse the same `take` trick here to clean this up:
```
let mut user_policy = self.user.take().unwrap_or_default();
while let Some(rule) = connect_rules.pop() {
    ...
}
self.user = Some(user_policy);
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+    /// - User context for `lorax` was removed
+    pub(crate) fn optimize(&mut self) {
+        // First, find the highest priority default connect rule, while removing all others
+        let priority_threshold: Option<u64> = if let Some(default_policy) = &mut self.default {
```
The priority calculation can be separated from the cleanup for improved readability:
```
        // First, find the highest priority default connect rule, while removing all others
        let priority_threshold: Option<u64> = self.default.as_ref().and_then(|default_policy| {
            let mut connect_rules: Vec<&Rule> = default_policy
                .rules
                .iter()
                .filter(|r| matches!(r, Rule::ConnectUser { .. }))
                .collect();
            connect_rules.sort_by(|a, b| b.get_priority().cmp(a.get_priority()));
            connect_rules.first().map(|r| *r.get_priority())
        });

       // Next, clean up all the default connect rules that don't match the highest priority rule
       if let (Some(priority), Some(policy)) = (priority_threshold, &mut self.default) {
           policy.rules.retain(|r| {
               if let Rule::ConnectUser { .. } = r {
                   *r.get_priority() == priority
               } else {
                   true
               }
           });
       }
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+                let mut priority = connect_rules
+                    .first()
+                    .map(|r| *r.get_priority())
+                    .unwrap_or(0);
```
This seems like it could be a helper function that's reused to find the default priority and to do the same for user connect rules.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+                    // Retain all rules except connect rules with lower priority
+                    // than the highest priority found
+                    !r.is_connect() || *r.get_priority() == priority
+                })
```
This is the same as the default policy cleanup pass, just "spelled" differently. Which is where the proposed refactor and helper functions would help, to express the same logical operation in the same way.

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+        } else {
+            base_policy.default = Some(default_rules);
+        }
+    }
```
Better variable names would help keep track of where we're working with the policy we just loaded, and where we're working with the base policy:
```suggestion
    if let Some(default_policy) = policy.default {
        if let Some(base_policy_default) = &mut base_policy.default {
            base_policy_default.rules.extend(default_policy.rules);
        } else {
            base_policy.default = Some(default_policy);
        }
    }
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,897 @@
+    }
+
+    if let Some(user_rules) = policy.user {
+        if let Some(user_context) = &mut base_policy.user {
```
as with the default policy:
```suggestion
    if let Some(user_policies) = policy.user {
        if let Some(base_policy_user) = &mut base_policy.user {
```

arnaldo2792 commented:
Forced push includes:
- Removes systemd drop-in and use an actual service instead

yeazelm commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,40 @@
+    # Allow replies
+    # This is a useless rule, it is dropped from the rules at runtime
+    # { send_type = "method_return", send_requested_reply = true, allow = true },
+    # This is a uselss rule, it is dropped from the rules at runtime
```
nit:
```suggestion
    # This is a useless rule, it is dropped from the rules at runtime
```

yeazelm commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,40 @@
+    # { send_type = "error", send_requested_reply = true, allow = true },
+    # Allow all receives
+    { receive_type = "method-call", allow = true },
+    # This is a userless rule, it is dropped from the rules at runtime
```
Is this supposed to be:
```suggestion
    # This is a useless rule, it is dropped from the rules at runtime
```

yeazelm commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,40 @@
+    { receive_type = "method-call", allow = true },
+    # This is a userless rule, it is dropped from the rules at runtime
+    # { receive_type = "method-return", allow = true },
+    # This is a userless rule, it is dropped from the rules at runtime
```
```suggestion
    # This is a useless rule, it is dropped from the rules at runtime
```

arnaldo2792 commented:
Forced push includes:
- Refactor to use `take` trick with optional values and other improvements (thanks @bcressey!)
- Align main documentation to <= 80 characters
- Fix epoch priority
- Wrap parts of error messages in `''`
- Remove useless comments and rules from `dbus-system.toml`
- Clarify comments for:
    - Wrapper around final policy
    - Launcher not hanging if the broker dies prior connection

bcressey commented on packages/os/dbus-1-system.toml:
```diff
@@ -0,0 +1,22 @@
+[user.root]
+rules = [
+    { send_destination = "org.freedesktop.DBus", send_interface = "org.freedesktop.systemd1.Activator", allow = true },
+    { send_destination = "org.freedesktop.DBus", send_interface = "org.freedesktop.DBus.Monitoring", allow = true },
+    { send_destination = "org.freedesktop.DBus", send_interface = "org.freedesktop.DBus.Debug.Stats", allow = true }
+]
```
nit: format these the same way as the rules in #678 for consistency.

bcressey commented on packages/os/whippet.service:
```diff
@@ -0,0 +1,21 @@
+ExecStart=/usr/bin/whippet
+
+[Install]
+Alias=dbus.service
```
Let's also alias to `dbus-broker.service` since it's still the broker at runtime.

bcressey commented on sources/whippet/src/main.rs:
```diff
@@ -0,0 +1,137 @@
+- Systemd socket activation
+
+## Writing D-Bus policies for whippet
+*whippet* uses TOML to define Dbus policies. This is an example of how to
```
```suggestion
*whippet* uses TOML to define D-Bus policies. This is an example of how to
```

bcressey commented on sources/whippet/src/dbus_policy.rs:
```diff
@@ -0,0 +1,441 @@
+            let mut batch: PolicyBatch = user_policy.try_into()?;
+            // If this policy batch didn't include a connect rule, use the default connect
+            // verdict and priority, matching the fallback behavior in the dbus-launcher
+            if !user_policy.has_connect_rule() {
+                batch.connect_verdict = default_batch.connect_verdict;
+                batch.connect_priority = default_batch.connect_priority;
+            }
```
nit: for consistency with `copy_rules_from`:
```suggestion
            if !user_policy.has_connect_rule() {
                batch.copy_connect_rule_from(&default_batch);
            }
```

bcressey commented on sources/whippet/src/policy.rs:
```diff
@@ -0,0 +1,923 @@
+        let policies_dir = policies_dir.as_ref();
+        let mut all_policies: Vec<PathBuf> = vec![policy_file.into()];
+
+        let mut toml_files: Vec<_> = fs::read_dir(policies_dir)
+            .context(error::ReadPathSnafu {
+                what: policies_dir.display().to_string(),
+            })?
+            .collect::<std::io::Result<Vec<_>>>()
+            .context(error::CollectPathsSnafu {
+                what: policies_dir.display().to_string(),
+            })?
+            .into_iter()
+            .map(|entry| entry.path())
+            .filter(|path| path.extension().and_then(|s| s.to_str()) == Some("toml"))
+            .collect();
```
nit: could factor out into a helper function to keep the main algorithm readable

yeazelm commented on sources/whippet/src/child.rs:
```diff
@@ -0,0 +1,128 @@
+                }
+                Ok(())
+            }
+            _ = sigterm.recv() => {
```
Do we care about SIGINT, can anything send other signals besides SIGTERM?

arnaldo2792 replied:
SIGINT is handled by `ctrl_c` above. And we don't care about other signals. The other signal `whippet` might care about in the future is `HUP` but only if we ever decide to support reloading the service without bringing the broker down.

arnaldo2792 commented:
Forced push includes:

- Fixes a bug that could cause the main process to hang. The broker could die before the it attempts to use the socket that it inherits, which will cause the call to send the policy to hang. Prevent this by spawning a thread to configure the broker, which will fail if the broker dies prior using the socket.

yeazelm commented on sources/whippet/src/broker.rs:
```diff
@@ -0,0 +1,155 @@
+    info!("Policy applied successfully");
+    notify_systemd_ready()?;
+    info!("Notified systemd that service is ready");
+    // The connection has to live for as long as whippet runs, otherwise the broker exits. Forget
+    // the connection to prevent dropping it
+    mem::forget(connection);
```
This feels like we may want a different approach? This does prevent the exit but should be instead be properly forking if we need to let this run outside of the context of the rest of the process?

bcressey replied:
> This feels like we may want a different approach? This does prevent the exit but should be instead be properly forking if we need to let this run outside of the context of the rest of the process?

I proposed `forget` - it's kind of brute force but it seems OK in this case, where we know that if either task exits then we'll be shutting down the program immediately, whereupon the kernel will run `Drop` for us 😀

yeazelm replied:
Ok, I think this makes sense. I can be good with this approach.

arnaldo2792 commented:
(forced push addresses commit message typo)

arnaldo2792 commented:
I confirmed that conformance testing passed for 1.34, aarch64, using whippet:

```
SUCCESS! -- 385 Passed
```
